tests/linen/module_test.py

```diff
@@ -590,6 +590,40 @@ def __call__(self, x):
     trace = mlp.apply(variables, x)
     self.assertEqual(trace, expected_trace)
 
+  def test_call_unbound_compact_module_methods(self):
+    dense = Dense(3)
+    with self.assertRaisesRegex(ValueError, "compact.*unbound module"):
+      dense(jnp.ones((1, )))
+
+
+  def test_call_unbound_noncompact_module_methods(self):
+    class EmptyModule(nn.Module):
+      foo: int = 3
+
+      def bar(self):
+        return self.foo
+
+    empty = EmptyModule()
+    # It's fine to call methods of unbound methods that don't depend on
+    # attributes defined during `setup`
+    self.assertEqual(empty.bar(), 3)
+
+
+  def test_call_unbound_noncompact_module_method_without_setup(self):
+    class EmptyModule(nn.Module):
+      def setup(self):
+        self.setup_called = True
+
+      def bar(self):
+        return self.setup_called
+
+    empty = EmptyModule()
```

Comment:
What about raising an error when `empty.setup()` is called directly ?

Reply:
Maybe as a separate change? Sounds like it's different than the issue in this PR? Want to file an issue or PR?

Reply:
Mentioning it here, because before this PR calling `empty.setup()` directly would have risen an exception:

`AssertionError: "compact.*unbound module" does not match "Can't call methods on orphaned modules"`

Once it's submitted I can create an issue if you prefer.

Reply:
Yes please file a new issue after this is submitted.